app/javascript/components/visualization/ScatterPlot.js

```diff
@@ -172,96 +191,204 @@ const ScatterPlot = withErrorBoundary(RawScatterPlot)
+  if (annotType === 'group' && !genes.length) {
+    const traceCounts = countOccurences(data.annotations)
+    trace.transforms = [{
+      type: 'groupby',
+      groups: data.annotations,
+      styles: annotValues.map((val, index) => {
+        return {
+          target: val,
+          value: {
+            name: `${val} (${traceCounts[val]} points)`,
+            marker: {
+              color: getColorBrewerColor(index),
+              size: pointSize
+            }
+          }
+        }
+      })
+    }]
+  } else {
+    trace.marker = {
+      line: { color: 'rgb(40,40,40)', width: 0 },
+      size: pointSize
+    }
+    const colors = genes.length ? data.expression : data.annotations
+    const title = genes.length ? axes.titles.magnitude : annotName
+    if (!isAnnotatedScatter) {
+      Object.assign(trace.marker, {
+        showscale: true,
+        colorscale: appliedScatterColor,
+        color: colors,
+        colorbar: { title, titleside: 'right' }
+      })
+      // if expression values are all zero, set max/min manually so the zeros still look like zero
+      // see SCP-2957
+      if (genes.length && !colors.some(val => val !== 0)) {
+        trace.marker.cmin = 0
+        trace.marker.cmax = 1
+      }
+    }
   }
+  addHoverLabel(trace, annotName, annotType, genes, isAnnotatedScatter, axes)
+  return [trace]
 }
 
-/** makes the data trace attributes (cells, trace name) available via hover text */
-function formatHoverLabels(data, annotationType, gene, isAnnotatedScatter) {
-  const groupHoverTemplate = '(%{x}, %{y})<br><b>%{text}</b><br>%{data.name}<extra></extra>'
-  data.forEach(trace => {
-    trace.text = trace.cells
-    if (!isAnnotatedScatter && (annotationType === 'numeric' || gene)) {
-      // use the 'meta' property so annotations are exposed to the hover template
-      // see https://community.plotly.com/t/hovertemplate-does-not-show-name-property/36139
-      trace.meta = trace.annotations
-      trace.hovertemplate = `(%{x}, %{y})<br>%{text} (%{meta})<br>
-        ${trace.marker.colorbar.title}: %{marker.color}<extra></extra>`
+/** return a hash of value=>count for the passed-in array
+    This i actually surprisingly quick even for large arrays, but we'd rather we
+    didn't have to do this.  See https://github.com/plotly/plotly.js/issues/5612s
+*/
+function countOccurences(array) {
+  return array.reduce((acc, curr) => {
+    if (!acc[curr]) {
+      acc[curr] = 1
     } else {
-      trace.text = trace.cells
-      trace.hovertemplate = groupHoverTemplate
+      acc[curr] += 1
     }
+    return acc
+  }, {})
+}
+
+/** sorts the scatter plot legend alphabetically
+  this is super-hacky in that it relies a lot on plotly internal classnames and styling implementation
+  it should only be used as a last resort if https://github.com/plotly/plotly.js/pull/5591
+  is not included in Plotly soon
+ */
+function sortLegend({ graphElementId, annotValues, isAnnotatedScatter, annotType, gene }) {
```

Comment:
Since the legend is auto-generated, and the sort order is determined by the input order of the traces (for group-based annotations), couldn't we just do a lexical sort of the traces by their name once we have all the data assembled?

Reply:
I wish -- that's what we used to do.  Now we only send a single trace to plotly, and plotly's groupBy transformation handles splitting it -- so we don't have access to the subsequent generated traces.  So plotly determines the order based on the order the annotations are encountered in the *individual cells*, which would be extremely difficult to reorder.